### PR TITLE
Completes implementation of divide & conquer method in STEDC

### PR DESCRIPTION
**Pull request by jzuniga-amd**
This PR also refactors STEDC unit test so that all the aspects of the divide & conquer algorithm (such as deflation) can be fully tested.

### CODE REVIEW

File: library/src/auxiliary/rocauxiliary_steqr.hpp [hunk 1]
```diff
@@ -330,7 +330,7 @@ void rocsolver_steqr_getMemorySize(const rocblas_evect evect,
     if(evect == rocblas_evect_none)
         *size_work_stack = sizeof(rocblas_int) * (2 * 32) * batch_count;
     else
-        *size_work_stack = sizeof(S) * (2 * n - 2) * batch_count;
+        *size_work_stack = sizeof(S) * (2 * n) * batch_count;
```
Comment (tfalders) [inline library/src/auxiliary/rocauxiliary_steqr.hpp:333]: Was there a memory access fault caused by this array being too small? Do we not also need to update `strideW` in `steqr_kernel`?

File: clients/include/testing_stedc.hpp [hunk 2]
```diff
@@ -438,7 +529,7 @@ void testing_stedc(Arguments& argus)
             if(argus.norm_check)
             {
                 rocsolver_bench_output("cpu_time_us", "gpu_time_us", "error");
-                rocsolver_bench_output(cpu_time_used, gpu_time_used, max_error);
+                rocsolver_bench_output(cpu_time_used, gpu_time_used, max_err);
```
Comment (tfalders) [inline clients/include/testing_stedc.hpp:532]: Should we not print `max(max_err, max_errv)`?

**Review by tfalders — COMMENTED** [review]
I still have to look over the changes in rocauxiliary_stedc.hpp, but here are my comments so far.

Also, there seems to be a CI failure in heevd on Centos 7.

**Review by qjojo — APPROVED** [review]
The algorithm is a bit beyond my reach but the glue looks good and I trust the tests. Nice job.

File: library/src/auxiliary/rocauxiliary_stedc.hpp [hunk 3]
```diff
@@ -15,79 +15,1308 @@
...
+        bb *= aa;
+        fx += bb;
+
+        // calculate tolerance er for convergence test
+        er += 8 * (hx - gx) + 2 * pinv + 3 * abs(bb) + abs(tau) * fdx;
+
+        // from now on, further step corrections will be calculated either with fixed weights method
+        // or with normal interpolation depending on the value of boolean fixed
+        cc = up ? -1 : 1;
+        fixed = (cc * fx > abs(oldfx) / 10) ? true : false;
```
Comment (cgmb) [inline library/src/auxiliary/rocauxiliary_stedc.hpp:278]: I would omit the redundant `? true : false`. IMO, it's easy enough to read something like

    fixed = (cc * fx) > (abs(oldfx) / 10);

File: library/src/auxiliary/rocauxiliary_stedc.hpp [hunk 4]
```diff
@@ -15,79 +15,1308 @@
...
+    }
+
+    *ev = x;
+    return converged ? 0 : 1;
+}
+
+/** STEDC_NUM_LEVS returns the ideal number of times/levels in which a matrix (or split block)
+    will be divided during the divide phase of divide & conquer algorithm.
+    i.e. number of sub-blocks = 2^levels **/
+__host__ __device__ inline rocblas_int stedc_num_levs(const rocblas_int n)
```
Comment (cgmb) [inline library/src/auxiliary/rocauxiliary_stedc.hpp:576]: Is the "lev" abbreviation coming from somewhere? I'm more familiar with "lvl". Though, given how long the function name is already, it seems like "level" would be fine too. It's only saving two characters after all.

File: library/src/auxiliary/rocauxiliary_stedc.hpp [hunk 5]
```diff
@@ -15,79 +15,1308 @@
...
+        levels = 8;
+    }
+    else if(n < 16)
+    {
+        levels = 0;
+    }
+    else
+    {
+        levels = n / 8;
+        levels = floor(log(levels) / log(2));
```
Comment (cgmb) [inline library/src/auxiliary/rocauxiliary_stedc.hpp:595]: I don't think this is in a hot path, so it doesn't really matter, but this pains me nonetheless. I've been meaning to ask about some of the solutions to this sort of thing... I suppose this is a good prompt to finally get around to it.

File: library/src/auxiliary/rocauxiliary_stedc.hpp [hunk 6]
```diff
@@ -15,79 +15,1308 @@
...
+                        }
+                        __syncthreads();
+                    }
+                    nrm = sqrt(nrm);
+
+                    // multiply by C (row by row)
+                    for(int ii = 0; ii < tsz; ++ii)
+                    {
+                        rocblas_int i = in + ii;
+                        go = go && (ii < sz);
```
Comment (cgmb) [inline library/src/auxiliary/rocauxiliary_stedc.hpp:1248]: Is the branch here intentional? The short-circuiting of logical operators introduces a branching point in this logic. If that's not intentional, I would probably suggest using the bitwise, `go &= (ii < sz)`.

**Review by cgmb — COMMENTED** [review]
Youch. I thought I was getting along well, but I clearly had more than two weeks remaining. Thanks so much for taking this over.

I have only minor comments.

File: library/src/auxiliary/rocauxiliary_steqr.hpp (hunk 1 above)
Comment (jzuniga-amd) [reply library/src/auxiliary/rocauxiliary_steqr.hpp:333]: I added the 2 extra elements just for simplicity when combining workspaces with stedc. 

Good catch! I corrected strideW.

File: library/src/auxiliary/rocauxiliary_stedc.hpp [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /************************************************************************
- * Derived from the BSD3-licensed
+  Derived from the BSD3-licensed
```
[the remark this answers is not in the document]
Comment (jzuniga-amd) [reply library/src/auxiliary/rocauxiliary_stedc.hpp:2]: I don't know. I did not touch this line.
I reversed the change. Let's see if the formatting removes the '*' again

File: library/src/auxiliary/rocauxiliary_stedc.hpp (hunk 5 above)
Comment (jzuniga-amd) [reply library/src/auxiliary/rocauxiliary_stedc.hpp:595]: Yeah, does not really matter. This function is only called once. Will leave it as it is for readability. Besides, this code will disappear once the tuning of the number of levels is done

File: clients/include/testing_stedc.hpp [hunk 8]
```diff
@@ -88,20 +88,107 @@ void stedc_initData(const rocblas_handle handle,
...
+            // The second sub-block will have some repeated eigenvalues in order to test the deflation process
+            S d;
+            rocblas_int NN1 = N1 / 2;
+            rocblas_int NN2 = N1 - NN1;
+            rocblas_int s1 = NN1 * NN1;
+            rocblas_int s2 = NN2 * NN2;
+            rocblas_int sw = NN2 * 32;
+            std::vector<S> A1(s1);
+            std::vector<S> A2(s2);
+            for(rocblas_int i = 0; i < NN1; ++i)
```
[the remark this answers is not in the document]
Comment (cgmb) [reply clients/include/testing_stedc.hpp:117]: Unfortunately, this is the iteration order used in the initialization loop in pretty much every test. I'm fine with fixing it on a case-by-case basis if @jzuniga-amd and @tfalders are. If not, I suppose we should finally create a ticket for fixing it across all the tests.

[the remark this answers is not in the document]
Comment (jzuniga-amd) [reply clients/include/testing_stedc.hpp:117]: Yeah, I have been waiting until we introduce openMP and other optimizations to the clients to take a look into these details across all tests... we have never had the time yet :(

[the remark this answers is not in the document]
Comment (cgmb) [reply clients/include/testing_stedc.hpp:117]: I was a bit worried that OpenMP would be a pain for packaging, but it's pretty easy to make it optional. rocSPARSE does it nicely, I think.

**Review by cgmb — APPROVED** [review]

**Review by tfalders — APPROVED** [review]
Unfortunately, my mind is too scattered today to give a thorough review of the algorithm, but what I've checked looks good!

File: library/src/auxiliary/rocauxiliary_stedc.hpp (hunk 5 above)
Comment (cgmb) [reply library/src/auxiliary/rocauxiliary_stedc.hpp:595]: For posterity, a better solution would be built on [`__builtin_clz`](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html). I linked the GCC documentation, but it's also available in clang.

**Pull request merged**